--- bzip2recover.c
+++ bzip2recover.c
@@ -36,8 +36,13 @@
    the 512MB limitation.
 */
 #ifdef __GNUC__
+# if defined(_LP64) || defined(__LP64__)
+   typedef  unsigned long long int  MaybeUInt64;
+#  define MaybeUInt64_FMT "%llu"
+# else
    typedef  unsigned long long int  MaybeUInt64;
 #  define MaybeUInt64_FMT "%Lu"
+# endif
 #else
 #ifdef _MSC_VER
    typedef  unsigned __int64  MaybeUInt64;
--- Makefile
+++ Makefile
@@ -47,7 +47,7 @@
       decompress.o \
       bzlib.o
 
-all: libbz2.a bzip2 bzip2recover test
+all: libbz2.a bzip2 bzip2recover
 
 bzip2: bzip2.o libbz2.a
 	$(CC) $(CPPFLAGS) $(BZIP_CFLAGS) $(LDFLAGS) $< -o $@ ./libbz2.a
@@ -66,7 +66,6 @@
 
 check: test
 test: bzip2
-	@cat words1
 	./bzip2 -1  < sample1.ref > sample1.rb2
 	./bzip2 -2  < sample2.ref > sample2.rb2
 	./bzip2 -3  < sample3.ref > sample3.rb2
@@ -79,7 +78,6 @@
 	cmp sample1.tst sample1.ref
 	cmp sample2.tst sample2.ref
 	cmp sample3.tst sample3.ref
-	@cat words3
 
 install: bzip2 bzip2recover
 	if ( test ! -d $(BINDIR) ) ; then mkdir -p $(BINDIR) ; fi
@@ -127,7 +125,6 @@
 
 # https://www.gnu.org/prep/standards/html_node/Makefile-Basics.html
 blocksort.o: blocksort.c
-	@cat words0
 	$(CC) $(CPPFLAGS) $(BZIP_CFLAGS) -c $< -o $@
 huffman.o: huffman.c
 	$(CC) $(CPPFLAGS) $(BZIP_CFLAGS) -c $< -o $@
@@ -171,10 +168,6 @@
 	   $(DISTNAME)/bzip2.1 \
 	   $(DISTNAME)/bzip2.1.preformatted \
 	   $(DISTNAME)/bzip2.txt \
-	   $(DISTNAME)/words0 \
-	   $(DISTNAME)/words1 \
-	   $(DISTNAME)/words2 \
-	   $(DISTNAME)/words3 \
 	   $(DISTNAME)/sample1.ref \
 	   $(DISTNAME)/sample2.ref \
 	   $(DISTNAME)/sample3.ref \
