--- libtextstyle/lib/libxml/timsort.h
+++ libtextstyle/lib/libxml/timsort.h
@@ -74,10 +74,6 @@
 static int compute_minrun(const uint64_t);
 
 #ifndef CLZ
-#ifdef __GNUC__
-#define CLZ __builtin_clzll
-#else
-
 static int clzll(uint64_t);
 
 /* adapted from Hacker's Delight */
@@ -124,7 +120,6 @@
 
 #define CLZ clzll
 #endif
-#endif
 
 static __inline int compute_minrun(const uint64_t size) {
   const int top_bit = 64 - CLZ(size);
--- gettext-tools/gnulib-lib/libxml/timsort.h
+++ gettext-tools/gnulib-lib/libxml/timsort.h
@@ -74,10 +74,6 @@
 static int compute_minrun(const uint64_t);
 
 #ifndef CLZ
-#ifdef __GNUC__
-#define CLZ __builtin_clzll
-#else
-
 static int clzll(uint64_t);
 
 /* adapted from Hacker's Delight */
@@ -124,7 +120,6 @@
 
 #define CLZ clzll
 #endif
-#endif
 
 static __inline int compute_minrun(const uint64_t size) {
   const int top_bit = 64 - CLZ(size);
--- gnulib-local/lib/libxml/timsort.h
+++ gnulib-local/lib/libxml/timsort.h
@@ -74,10 +74,6 @@
 static int compute_minrun(const uint64_t);
 
 #ifndef CLZ
-#ifdef __GNUC__
-#define CLZ __builtin_clzll
-#else
-
 static int clzll(uint64_t);
 
 /* adapted from Hacker's Delight */
@@ -124,7 +120,6 @@
 
 #define CLZ clzll
 #endif
-#endif
 
 static __inline int compute_minrun(const uint64_t size) {
   const int top_bit = 64 - CLZ(size);
