--- lib/utf8mac.h
+++ lib/utf8mac.h
@@ -37,10 +37,6 @@
 # else
 #  define OSSwapInt16(x) ((((x) & 0xff) << 8) | (((x) >> 8) & 0xff))
 # endif
-  typedef uint8_t u_int8_t;
-  typedef uint16_t u_int16_t;
-  typedef uint32_t u_int32_t;
-  typedef uint64_t u_int64_t;
 #elif defined(_AIX)
 # include <inttypes.h>
 # define OSSwapInt16(x) ((((x) & 0xff) << 8) | (((x) >> 8) & 0xff))
@@ -48,6 +44,12 @@
 # include <byteswap.h>
 # define OSSwapInt16(x) bswap_16(x)
 #endif /* __sun */
+
+/* https://github.com/fumiyas/libiconv-utf8mac/issues/3 */
+typedef uint8_t u_int8_t;
+typedef uint16_t u_int16_t;
+typedef uint32_t u_int32_t;
+typedef uint64_t u_int64_t;
 
 #ifdef WORDS_LITTLEENDIAN
 # define __LITTLE_ENDIAN__
--- lib/utf8mac.h
+++ lib/utf8mac.h
@@ -27,6 +27,13 @@
  	Includes Unicode 3.2 decomposition code derived from Core Foundation
  */
 
+#ifdef HAVE_STDINT_H
+# include <stdint.h>
+#endif
+#ifdef HAVE_INTTYPES_H
+# include <inttypes.h>
+#endif
+
 #if defined(__APPLE__)
 #include <libkern/OSByteOrder.h>
 #elif defined(__sun)
