--- xts.c
+++ xts.c
@@ -49,15 +49,28 @@
 /* the algorithm reads this as a 128bit Little Endian number */
 /* src and dest can point to the same buffer for in-place operations */
 #if WORDS_BIGENDIAN
+
+#if (_LP64 || __LP64__)
 #define BE_SHIFT(x) ((((x) & 0x7f7f7f7f7f7f7f7f) << 1) | \
                      (((x) & 0x8080808080808080) >> 15))
+#else
+#define BE_SHIFT(x) ((((x) & 0x7f7f7f7f7f7f7f7full) << 1) | \
+                     (((x) & 0x8080808080808080ull) >> 15))
+#endif
+
 static void
 xts_shift(union nettle_block16 *dst,
           const union nettle_block16 *src)
 {
+#if (_LP64 || __LP64__)
   uint64_t carry = (src->u64[1] & 0x80) >> 7;
   dst->u64[1] = BE_SHIFT(src->u64[1]) | ((src->u64[0] & 0x80) << 49);
   dst->u64[0] = BE_SHIFT(src->u64[0]) ^ (0x8700000000000000 & -carry);
+#else
+  uint64_t carry = (src->u64[1] & 0x80) >> 7;
+  dst->u64[1] = BE_SHIFT(src->u64[1]) | ((src->u64[0] & 0x80) << 49);
+  dst->u64[0] = BE_SHIFT(src->u64[0]) ^ (0x8700000000000000ull & -carry);
+#endif
 }
 #else /* !WORDS_BIGENDIAN */
 static void
--- ctr.c
+++ ctr.c
@@ -48,6 +48,10 @@
 
 #define MIN(a,b) (((a) < (b)) ? (a) : (b))
 
+#if 1
+# undef HAVE_BUILTIN_BSWAP64
+#endif
+
 static size_t
 ctr_fill (size_t block_size, uint8_t *ctr, size_t length, uint8_t *buffer)
 {
--- testsuite/Makefile.in
+++ testsuite/Makefile.in
@@ -138,7 +138,6 @@
 
 # The PATH update is for windows dlls, DYLD_LIBRARY_PATH is for OSX.
 check: $(TS_ALL)
-	LD_LIBRARY_PATH=../.lib PATH="../.lib:$$PATH" DYLD_LIBRARY_PATH=../.lib \
 	  srcdir="$(srcdir)" \
 	  EMULATOR="$(EMULATOR)" NM="$(NM)" EXEEXT="$(EXEEXT)" \
           $(top_srcdir)/run-tests $(TS_ALL)
--- examples/Makefile.in
+++ examples/Makefile.in
@@ -115,7 +115,6 @@
 
 # The PATH update is for windows dlls, DYLD_LIBRARY_PATH is for OSX.
 check: $(TS_ALL)
-	LD_LIBRARY_PATH=../.lib PATH="../.lib:$$PATH" DYLD_LIBRARY_PATH=../.lib \
 	  srcdir="$(srcdir)" EMULATOR="$(EMULATOR)" EXEEXT="$(EXEEXT)" \
           "$(top_srcdir)"/run-tests $(TS_ALL)
 
--- run-tests
+++ run-tests
@@ -28,6 +28,13 @@
 
 export srcdir
 
+nettle_lib_dir=`dirname $PWD`/.lib
+LD_LIBRARY_PATH=$nettle_lib_dir
+DYLD_LIBRARY_PATH=$nettle_lib_dir
+
+export LD_LIBRARY_PATH
+export DYLD_LIBRARY_PATH
+
 # When used in make rules, we sometimes get the filenames VPATH
 # expanded, but usually not.
 find_program () {
