--- p11-kit/rpc-message.c
+++ p11-kit/rpc-message.c
@@ -577,7 +577,7 @@
 uint16_t
 p11_rpc_buffer_decode_uint16 (unsigned char* data)
 {
-	uint16_t value = data[0] << 8 | data[1];
+	uint16_t value = ((uint16_t)data[0]) << 8 | data[1];
 	return value;
 }
 
@@ -636,7 +636,7 @@
 uint32_t
 p11_rpc_buffer_decode_uint32 (unsigned char* ptr)
 {
-	uint32_t val = ptr[0] << 24 | ptr[1] << 16 | ptr[2] << 8 | ptr[3];
+	uint32_t val = ((uint32_t)ptr[0]) << 24 | ((uint32_t)ptr[1]) << 16 | ((uint32_t)ptr[2]) << 8 | ptr[3];
 	return val;
 }
 

--- trust/utf8.c
+++ trust/utf8.c
@@ -147,7 +147,7 @@
 			 */
 			return -1;
 		}
-		uch <<= 6;
+		uch = ((uint32_t)uch) << 6;
 		uch |= *str++ & 0x3f;
 	}
 	if (uch < lbound) {
@@ -237,7 +237,7 @@
 	if (len < 2)
 		return -1;
 
-	*wc = (str[0] << 8 | str[1]);
+	*wc = (((uint32_t)str[0]) << 8 | str[1]);
 	return 2;
 }
 
@@ -253,7 +253,7 @@
 	if (len < 4)
 		return -1;
 
-	*uc = (str[0] << 24 | str[1] << 16 | str[2] << 8 | str[3]);
+	*uc = (((uint32_t)str[0]) << 24 | ((uint32_t)str[1]) << 16 | ((uint32_t)str[2]) << 8 | str[3]);
 	return 4;
 }
 
