--- Makefile.in
+++ Makefile.in
@@ -17,11 +17,14 @@
 
 #### Start of system configuration section. ####
 
+SHELL = /bin/sh
+
 srcdir = @srcdir@
 VPATH = @srcdir@
 
 CC = @CC@
 AR = ar
+ARFLAGS = @ARFLAGS@
 RANLIB = @RANLIB@
 
 INSTALL = @INSTALL@
@@ -31,7 +34,7 @@
 
 DEFS = @DEFS@ -DTERMCAP_FILE=\"$(termcapfile)\"
 
-CFLAGS = -g
+CFLAGS = @CFLAGS@
 
 prefix = @prefix@
 exec_prefix = @exec_prefix@
@@ -42,12 +45,6 @@
 # Directory in which to install termcap.h.
 includedir = $(prefix)/include
 
-# Directory in which to optionally also install termcap.h,
-# so compilers besides gcc can find it by default.
-# If it is empty or not defined, termcap.h will only be installed in
-# includedir. 
-oldincludedir = /usr/include
-
 # Directory in which to install the documentation info files.
 infodir = $(prefix)/info
 
@@ -57,8 +54,6 @@
 
 #### End of system configuration section. ####
 
-SHELL = /bin/sh
-
 SRCS = termcap.c tparam.c version.c
 OBJS = termcap.o tparam.o version.o
 HDRS = termcap.h
@@ -69,21 +64,16 @@
 all:	libtermcap.a info
 
 .c.o:
-	$(CC) -c $(CPPFLAGS) $(DEFS) -I. -I$(srcdir) $(CFLAGS) $<
+	$(CC) $(CPPFLAGS) $(CFLAGS) $(DEFS) -I. -I$(srcdir) -c $< -o $@
 
 install: all installdirs @installdata@
 	$(INSTALL_DATA) libtermcap.a $(libdir)/libtermcap.a
-	-$(RANLIB) $(libdir)/libtermcap.a
 	cd $(srcdir); $(INSTALL_DATA) termcap.h $(includedir)/termcap.h
-	-cd $(srcdir); test -z "$(oldincludedir)" || \
-	  $(INSTALL_DATA) termcap.h $(oldincludedir)/termcap.h
-	cd $(srcdir); for f in termcap.info*; \
-	do $(INSTALL_DATA) $$f $(infodir)/$$f; done
+	-cd $(srcdir); test -z "$(includedir)" || \
+	  $(INSTALL_DATA) termcap.h $(includedir)/termcap.h
 
 uninstall: @uninstalldata@
 	rm -f $(libdir)/libtermcap.a $(includedir)/termcap.h
-	test -z "$(oldincludedir)" || rm -f $(oldincludedir)/termcap.h
-	rm -f $(infodir)/termcap.info*
 
 # These are separate targets to avoid trashing the user's existing
 # termcap file unexpectedly.
--- tparam.c
+++ tparam.c
@@ -40,10 +40,8 @@
 
 #endif /* not emacs */
 
-#ifndef NULL
-#define NULL (char *) 0
-#endif
-
+#include <unistd.h>
+
 #ifndef emacs
 static void
 memory_out ()
@@ -75,7 +73,7 @@
   return tem;
 }
 #endif /* not emacs */
-
+
 /* Assuming STRING is the value of a termcap string entry
    containing `%' constructs to expand parameters,
    merge in parameter values and store result in block OUTSTRING points to.
@@ -165,7 +163,7 @@
 	      outlen *= 2;
 	      new = (char *) xrealloc (outstring, outlen);
 	    }
-	  
+
 	  op = new + offset;
 	  outend = new + outlen;
 	  outstring = new;
@@ -312,7 +310,7 @@
     strcat (op, left);
   return outstring;
 }
-
+
 #ifdef DEBUG
 
 main (argc, argv)
