--- bootstrap
+++ bootstrap
@@ -254,6 +254,13 @@
       # presented with env variables to set to find the right version
       ($2 --version </dev/null)
     fi
+  elif test "$1" = "-V"; then
+    ($2 -V </dev/null) >/dev/null 2>&1
+    if test $? -ge 126; then
+      # If not found, run with diagnostics as one may be
+      # presented with env variables to set to find the right version
+      ($2 -V </dev/null)
+    fi
   else
     ($1 --version </dev/null) >/dev/null 2>&1
   fi
